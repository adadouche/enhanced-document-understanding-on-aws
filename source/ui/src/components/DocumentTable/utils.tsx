/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import { useEffect, useState } from 'react';

export const EMPTY_PANEL_CONTENT = {
    header: 'No case selected',
    body: 'Select a case to see its documents.'
};

export const useSplitPanel = (selectedItems: any) => {
    const [splitPanelSize, setSplitPanelSize] = useState(500);
    const [splitPanelOpen, setSplitPanelOpen] = useState(false);
    const [hasManuallyClosedOnce, setHasManuallyClosedOnce] = useState(false);

    const onSplitPanelResize = ({ detail: { size } }: any) => {
        setSplitPanelSize(size);
    };

    const onSplitPanelToggle = ({ detail: { open } }: any) => {
        setSplitPanelOpen(open);

        if (!open) {
            setHasManuallyClosedOnce(true);
        }
    };

    useEffect(() => {
        if (selectedItems.length && !hasManuallyClosedOnce) {
            setSplitPanelOpen(true);
        }
    }, [selectedItems.length, hasManuallyClosedOnce]);

    return {
        splitPanelOpen,
        onSplitPanelToggle,
        splitPanelSize,
        onSplitPanelResize
    };
};
