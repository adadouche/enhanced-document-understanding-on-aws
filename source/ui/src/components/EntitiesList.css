.ObjectDisplay {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
  
  .ColumnHeader {
    font-weight: bold;
  }
  
  .ObjectRow {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }
  
  .ObjectCell {
    margin: 5px;
  }
  