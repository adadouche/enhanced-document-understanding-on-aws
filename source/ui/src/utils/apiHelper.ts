// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Auth } from "aws-amplify";

let runtimeConfig: any | undefined;

/**
 *
 */
export async function generateToken() {
    try {
        const user = await Auth.currentAuthenticatedUser();
        const token = user.getSignInUserSession().getIdToken().getJwtToken();
        return token;
    } catch (error) {
        console.error('error REST API:', error);
    }
}

/**
 * Gets token from cognito
 */

export const getRuntimeConfig = async () => {
    if(runtimeConfig === undefined) {
        runtimeConfig = (await fetch("/runtimeConfig.json")).json();
    }
    return await runtimeConfig;
};

export async function signOut() {
    await Auth.signOut();
}
