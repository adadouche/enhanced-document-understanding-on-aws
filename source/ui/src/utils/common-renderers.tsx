/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the 'License'). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import StatusIndicator from '@cloudscape-design/components/status-indicator';

/**
* Enables conditional rendering of StatusIndicator whenever there are error/loading statuses.
* Returns null for 'success' to enable rendering content fetched from the API.
*
* @param status - the status that is to be rendered as a part of StatusIndicator
* @param showStatusLabel - boolean flag to enable message rendering along with the status
* @param showSuccessMessage - boolean flag to show if the success state has to be rendered.
    This can be set to false for rendering user content.
* @param errorMessage - error message that should be displayed when the currentStatus is in error state
*/
export const renderStatus = function (
    status: any,
    showStatusLabel: any,
    showSuccessMessage: any,
    errorMessage: any,
    successMessage: any
) {
    let statusLabel = '';
    let statusIndicatorComponent = null;
    errorMessage = errorMessage ? errorMessage : 'An error occurred.';

    if (status === 'error') {
        statusLabel = showStatusLabel ? errorMessage : '';
        statusIndicatorComponent = <StatusIndicator type={status}>{statusLabel}</StatusIndicator>;
    } else if (status === 'loading') {
        statusLabel = showStatusLabel ? 'Loading' : '';
        statusIndicatorComponent = <StatusIndicator type={status}>{statusLabel}</StatusIndicator>;
    } else if (status === 'success' && showSuccessMessage) {
        successMessage = successMessage ? successMessage : 'Success';
        statusLabel = showStatusLabel ? successMessage : '';
        statusIndicatorComponent = <StatusIndicator type={status}>{statusLabel}</StatusIndicator>;
    }
    return statusIndicatorComponent;
};

/**
 * Enables testing for successful status when rendering html for conditional rendering
 *
 * @param status - the status that is to be rendered as a part of StatusIndicator
 */
export const isStatusSuccess = function (status: any) {
    let isSuccess = status === 'success' ? true : false;
    return isSuccess;
};
