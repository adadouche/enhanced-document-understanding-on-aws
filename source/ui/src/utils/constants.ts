/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

export const API_NAME = 'api';

export const COMPREHEND_SERVICE = 'COMPREHEND';
export const COMPREHEND_MEDICAL_SERVICE = 'COMPREHEND_MEDICAL';
export const TEXTRACT_RAW_TEXT = 'TEXTRACT_RAW_TEXT';
export const TEXTRACT_KEY_VALUE_PAIRS = 'TEXTRACT_KEY_VALUE_PAIRS';
export const TEXTRACT_TABLES = 'TEXTRACT_TABLES';
export const MIN_SEARCH_QUERY_LENGTH = 1;
export const CASE_PLACEHOLDER_DOCUMENT_ID = '0000';
export const MIN_CASE_NAME_LENGTH = 3;
export const MAX_CASE_NAME_LENGTH = 50;
export const PREVIEW_REDACTION_ON = 'PREVIEW_REDACTION_ON';
export const ENTITIES = 'entities';
export const MAX_UPLOAD_FILE_SIZE = 5000000;

export const EntityTypes = {
    ENTITY_STANDARD: 'entity-standard',
    PII: 'entity-pii',
    MEDICAL_ENTITY: 'entity-medical'
};

export enum InferenceName  {
    TEXTRACT_DETECT_TEXT= 'textract-detectText',
    TEXTRACT_ANALYZE_TEXT= 'textract-analyzeDoc',
    COMPREHEND_GENERIC= 'entity-standard-locations',
    COMPREHEND_PII= 'entity-pii-locations',
    COMPREHEND_MEDICAL= 'entity-medical-locations'
};

export const FacetDocumentAttributeKey = {
    'file_type': 'File type',
    'doc_type': 'Document type'
};

/**
 * Maps case status from backend to a friendly string for display
 */
export const CaseStatusDisplayTextMapping: Record<string, string> = {
    'initiate': 'Created',
    'in-process': 'Processing',
    'failure': 'Failed',
    'success': 'Complete'
};

export const DISPLAY_DATE_FORMAT = 'DD-MMM-YYYY, hh:mm:ss A';
