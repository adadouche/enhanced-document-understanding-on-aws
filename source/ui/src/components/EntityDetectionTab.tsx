/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Box, Container, StatusIndicatorProps } from '@cloudscape-design/components';
import { useCallback, useMemo } from 'react';
import { renderStatus } from '../utils/common-renderers';
import { BoundingBox } from '../utils/interfaces';

import { EntityTypes } from '../utils/constants';
import DocumentRenderer from './DocumentRenderer/DocumentRenderer';
import EntitiesList from './EntitiesList';
import { getEntitiesToProcess } from './entityUtils';

type EntityDetectionTabProps = {
    selectedDocumentFileType: string | null;
    selectedDocumentUrl: string | null;
    selected?: string;
    standardEntities: any;
    medicalEntities: any;
    piiEntities: any;
    documentPageCount: number;
    currentPageNumber: number;
    switchPage: Function;
    comprehendService: string;
    entityType: string;
    selectedDocumentId: string | null;
    selectedCaseId: string | null;
    selectedEntities: any;
    setSelectedEntities: Function;
    previewRedaction: string;
    setPreviewRedaction: Function;
    currentStatus: StatusIndicatorProps.Type | undefined;
    dataTestId?: string;
    retrieveSignedUrl: Function;
};

/**
 * EntityDetectionTab is used to render data for Generic, PII and Medical Comprehend inferences
 * @param props to be used to populate the tab
 * @returns
 */
export default function EntityDetectionTab(props: EntityDetectionTabProps) {
    let documentEntities = getEntitiesToProcess(props.entityType, props);

    const getFilteredArray = useCallback(
        (entityType: string) => {
            return props.selectedEntities[entityType].filter((item: string[]) => {
                if (item.length > 2 && item[2] !== props.currentPageNumber.toString()) {
                    return false;
                }

                if (
                    props.selectedEntities[entityType].some(
                        (otherItem: string[]) =>
                            otherItem.length < item.length &&
                            otherItem.slice(0, item.length - 1).every((val, index) => val === item[index])
                    )
                ) {
                    return false;
                }

                return true;
            });
        },
        [props.currentPageNumber, props.selectedEntities]
    );

    const getBoundingBoxesForEntityPage = useCallback((entities: any, entityPath: string[]) => {
        let boundingBoxes: BoundingBox[] = [];
        for (const pageInstance of entities[entityPath[0]][entityPath[1]][entityPath[2]]) {
            boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
        }
        return boundingBoxes;
    }, []);

    const getBoundingBoxesForEntityValue = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            if (props.currentPageNumber.toString() in entities[entityPath[0]][entityPath[1]]) {
                for (const pageInstance of entities[entityPath[0]][entityPath[1]][props.currentPageNumber.toString()]) {
                    boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const getBoundingBoxesForEntityType = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            for (const entity of Object.keys(entities[entityPath[0]])) {
                if (props.currentPageNumber.toString() in entities[entityPath[0]][entity]) {
                    for (const pageInstance of entities[entityPath[0]][entity][props.currentPageNumber.toString()]) {
                        boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                    }
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const pageEntities = useMemo(() => {
        let boundingBoxes: BoundingBox[] = [];
        for (const entityType of Object.values(EntityTypes)) {
            if (props.selectedEntities[entityType] && props.selectedEntities[entityType].length > 0) {
                const entities = getEntitiesToProcess(entityType, props);
                const filteredArray = getFilteredArray(entityType);
                for (const entityPath of filteredArray) {
                    switch (entityPath.length) {
                        case 3:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityPage(entities, entityPath));
                            break;
                        case 2:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityValue(entities, entityPath));
                            break;
                        case 1:
                        default:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityType(entities, entityPath));
                            break;
                    }
                }
            }
        }

        return boundingBoxes;
    }, [
        getBoundingBoxesForEntityPage,
        getBoundingBoxesForEntityType,
        getBoundingBoxesForEntityValue,
        getFilteredArray,
        props.medicalEntities,
        props.piiEntities,
        props.selectedEntities,
        props.standardEntities
    ]);

    const status = renderStatus(
        props.currentStatus,
        true,
        false,
        `An error occurred loading the document preview.`,
        ''
    );
    return (
        <div style={{ display: 'flex', height: '100%' }} data-testid={props.dataTestId}>
            <div style={{ width: '50%', float: 'left', paddingRight: '0.5%', paddingLeft: '1%' }}>
                <Container data-testid="document-rendering-container">
                    {status}
                    <DocumentRenderer
                        selectedDocumentFileType={props.selectedDocumentFileType}
                        selectedDocumentUrl={props.selectedDocumentUrl}
                        currentPageNumber={props.currentPageNumber}
                        switchPage={props.switchPage}
                        marks={pageEntities}
                        previewRedaction={props.previewRedaction}
                        retrieveSignedUrl={props.retrieveSignedUrl}
                    />
                </Container>
            </div>
            <div
                style={{
                    width: '50%',
                    float: 'left',
                    paddingLeft: '0.5%',
                    paddingRight: '1%',
                    height: '100%'
                }}
            >
                <Container fitHeight={true}>
                    <div>
                        <Box data-testid="tab-box">
                            <EntitiesList
                                entities={documentEntities}
                                documentPageCount={props.documentPageCount}
                                currentPageNumber={props.currentPageNumber}
                                switchPage={props.switchPage}
                                comprehendService={props.comprehendService}
                                entityType={props.entityType}
                                standardEntities={props.standardEntities}
                                medicalEntities={props.medicalEntities}
                                piiEntities={props.piiEntities}
                                selectedEntities={props.selectedEntities}
                                setSelectedEntities={props.setSelectedEntities}
                                selectedDocumentId={props.selectedDocumentId}
                                selectedCaseId={props.selectedCaseId}
                                previewRedaction={props.previewRedaction}
                                setPreviewRedaction={props.setPreviewRedaction}
                                currentStatus={props.currentStatus}
                            />
                        </Box>
                    </div>
                </Container>
            </div>
        </div>
    );
}
