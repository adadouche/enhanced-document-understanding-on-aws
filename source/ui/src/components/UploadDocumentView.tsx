/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the License). You may not use this file except in compliance      *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/
import {
    Alert,
    AppLayout,
    Button,
    Container,
    ContentLayout,
    Form,
    FormField,
    Header,
    ProgressBar,
    Select,
    SelectProps,
    SpaceBetween
} from '@cloudscape-design/components';
import { generateToken, getUsername } from './DocumentTable/DocumentTable';
import { FileSize, FileType } from './FileUpload/interfaces';

import { API } from 'aws-amplify';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { API_NAME, MAX_UPLOAD_FILE_SIZE } from '../utils/constants';
import FileUpload from './FileUpload/FileUpload';
import { mapResultsToCases } from './makeData';
import { formatFileSize } from './FileUpload/internal';

export interface WorkflowConfig {
    NumRequiredDocuments: number;
    UniqueDocumentTypes: string[];
    WorkflowConfigName: string;
}

type UploadDocumentViewProps = {
    caseName?: string;
    caseId?: string;
    casesList: any[];
    setCasesList: Function;
    workflowConfig: WorkflowConfig;
    selectedCaseRemainingRequiredDocs: any;
};

type Case = {
    name: string;
    status: string;
    caseDocuments: Object[];
    dateCreated: string;
};

type CaseMap = {
    [key: string]: Case;
};

let token: string;
let results: any;

enum Status {
    IN_PROGRESS = 'in-progress',
    SUCCESS = 'success',
    ERROR = 'error'
}

export default function UploadDocumentView(props: UploadDocumentViewProps) {
    let caseMap: CaseMap = {};
    let casesOptions: SelectProps.Option[] = [];

    const updateCaseMap = () => {
        if (props.casesList) {
            for (const item of props.casesList) {
                const _case: Case = {
                    name: item.name,
                    status: item.status,
                    caseDocuments: item.caseDocuments,
                    dateCreated: item.dateCreated
                };
                caseMap[item.caseId] = _case;
            }
        }
        for (const caseId in caseMap) {
            casesOptions.push({
                label: caseMap[caseId].name,
                value: caseId,
                description: `Case Id: ${caseId}`,
                disabled:
                    caseMap[caseId].status !== 'initiate' ||
                    caseMap[caseId].caseDocuments.length >= props.workflowConfig.NumRequiredDocuments
            });
        }
    };

    updateCaseMap();

    let caseId: string | undefined;
    if (props.caseId) {
        caseId = props.caseId;
    }

    const [initialSelectedCaseId] = React.useState<string>(caseId!);
    const [initialSelectedCaseName] = React.useState<string>(caseMap[initialSelectedCaseId]?.name);
    const [selectedCaseOption, setSelectedCaseOption] = React.useState<SelectProps.Option>({
        label: initialSelectedCaseName,
        value: initialSelectedCaseId
    });
    const [chosenFiles, setChosenFiles] = React.useState<File[]>([]);
    const [showProgressBar, setShowProgressBar] = React.useState(false);
    const [progressStatus, setProgressStatus] = React.useState<Status | undefined>();
    const [filesFailedToUpload, setFilesFailedToUpload] = React.useState<string[]>([]);
    const [uploadResultMessage, setUploadResultMessage] = React.useState('');
    const [uploadProgress, setUploadProgress] = React.useState(0);
    const [chooseFilesError, setChooseFilesError] = React.useState('');
    const [allFilesAttempted, setAllFilesAttempted] = React.useState(false);
    const [documentTypeList, setDocumentTypeList] = React.useState<string[]>([]);
    const [refreshCaseData, setRefreshCaseData] = React.useState(false);
    const [selectedCaseRemainingAllowedDocs, setSelectedCaseRemainingAllowedDocs] = React.useState();
    const navigate = useNavigate();

    const fileMetadata = { name: true, type: true, size: FileSize.MB, lastModified: true, thumbnail: true };
    const allowMultipleFiles = true;

    React.useEffect(() => {
        const newRemainingDocs = Object.assign({}, props.selectedCaseRemainingRequiredDocs);

        for (let docType of documentTypeList) {
            if (newRemainingDocs[docType]) {
                newRemainingDocs[docType] -= 1;
            }
        }
        setSelectedCaseRemainingAllowedDocs(newRemainingDocs);
    }, [documentTypeList, props.selectedCaseRemainingRequiredDocs]);

    React.useEffect(() => {
        const fetchData = async () => {
            token = await generateToken();
            results = await API.get(API_NAME, 'cases', {
                headers: {
                    Authorization: token
                }
            });
            const newData = mapResultsToCases(results.Items);
            props.setCasesList(newData);
        };
        if (!props.casesList || refreshCaseData) {
            fetchData();
            updateCaseMap();
            setRefreshCaseData(false);
        }
    }, [refreshCaseData]);

    const handleFileSelection = async (newFiles: FileType) => {
        setChooseFilesError('');
        if (newFiles instanceof File) {
            if (MAX_UPLOAD_FILE_SIZE && newFiles.size > MAX_UPLOAD_FILE_SIZE) {
                setChooseFilesError(
                    'File size must be less than ' + formatFileSize(MAX_UPLOAD_FILE_SIZE, fileMetadata)
                );
            } else {
                setChosenFiles([newFiles]);
            }
        } else if (newFiles instanceof Array) {
            if (MAX_UPLOAD_FILE_SIZE) {
                const numFilesAttempted = newFiles.length;
                newFiles = newFiles.filter((file) => file.size <= MAX_UPLOAD_FILE_SIZE);
                if (newFiles.length < numFilesAttempted) {
                    setChooseFilesError(
                        'File size must be less than ' + formatFileSize(MAX_UPLOAD_FILE_SIZE, fileMetadata)
                    );
                }
            }
            setChosenFiles(newFiles);
        }
    };

    const uploadFile = async (
        files: File[],
        index: number,
        postPolicyFields: { [key: string]: string },
        bucketUrl: string
    ) => {
        let postData = new FormData();
        for (const [key, entry] of Object.entries(postPolicyFields)) {
            postData.append(key, entry);
        }
        postData.append('file', files[index]);
        const data = {
            method: 'POST',
            body: postData
        };
        const response = await fetch(bucketUrl, data);
        if (200 <= response.status && response.status < 300) {
            setUploadProgress((uploadProgress) => uploadProgress + Math.ceil((1 / chosenFiles.length) * 100));
        } else {
            setFilesFailedToUpload((filesFailedToUpload) => [...filesFailedToUpload, files[index].name]);
        }
        if (index === chosenFiles.length - 1) {
            setAllFilesAttempted(true);
        }
    };

    const postResource = async (endpoint: string, params = {}) => {
        try {
            token = await generateToken();
            const response = await API.post(API_NAME, endpoint, {
                headers: {
                    Authorization: token
                },
                body: params
            });
            return response;
        } catch (err) {
            console.error(err);
        }
    };

    const handleUploadClick = async () => {
        setProgressStatus(Status.IN_PROGRESS);
        setFilesFailedToUpload((filesFailedToUpload) => []);
        setUploadResultMessage('');
        setUploadProgress(0);
        setShowProgressBar(true);
        setAllFilesAttempted(false);

        const userName = await getUsername();
        let promiseArray = [];
        for (let i = 0; i < chosenFiles.length; i++) {
            let apiRequest = postResource(`document`, {
                userId: userName,
                caseId: selectedCaseOption.value,
                caseName: selectedCaseOption.label,
                fileName: chosenFiles[i].name,
                fileExtension: '.' + chosenFiles[i].name.split('.').pop(),
                documentType: documentTypeList[i],
                tagging: `<Tagging><TagSet><Tag><Key>userId</Key><Value>${userName}</Value></Tag></TagSet></Tagging>`
            });
            promiseArray.push(apiRequest);
        }

        await Promise.allSettled(promiseArray).then(async (results) => {
            results.forEach((result, i) => {
                if (result.status === 'fulfilled') {
                    if (result.value === undefined) {
                        setFilesFailedToUpload((filesFailedToUpload) => [...filesFailedToUpload, chosenFiles[i].name]);
                        setUploadResultMessage((uploadResultMessage) => 'File upload failed');
                        setProgressStatus(Status.ERROR);
                    } else {
                        uploadFile(chosenFiles, i, result.value.fields, result.value.url);
                    }
                }
            });
        });
        setRefreshCaseData(true);
        setChosenFiles([]);
    };

    const allowChooseFiles = () => {
        if (selectedCaseOption.value !== undefined && caseMap[selectedCaseOption.value] !== undefined) {
            return (
                chosenFiles.length <
                props.workflowConfig.NumRequiredDocuments - caseMap[selectedCaseOption.value].caseDocuments.length
            );
        }
        return false;
    };

    const updateDocumentTypeList = (index: number, value: string | null) => {
        const updatedList = [...documentTypeList];
        if (index >= updatedList.length) {
            updatedList.push(value!);
        } else if (value === null) {
            updatedList.splice(index, 1);
        } else {
            updatedList[index] = value;
        }
        setDocumentTypeList(updatedList);
    };

    const allowUploadFiles = () => {
        if (selectedCaseOption.value !== undefined && caseMap[selectedCaseOption.value] !== undefined) {
            return (
                chosenFiles.length > 0 &&
                caseMap[selectedCaseOption.value].caseDocuments.length + chosenFiles.length <=
                    props.workflowConfig.NumRequiredDocuments
            );
        }
        return false;
    };

    React.useEffect(() => {
        if (uploadProgress < 100 && allFilesAttempted) {
            setProgressStatus(Status.ERROR);
            setUploadResultMessage((uploadResultMessage) => 'File upload failed');
        } else if (uploadProgress >= 100 && allFilesAttempted) {
            setProgressStatus(Status.SUCCESS);
            setUploadResultMessage((uploadResultMessage) => 'Files uploaded sucessfully');
        } else if (uploadProgress > 0 && uploadProgress < 100 && !allFilesAttempted) {
            setProgressStatus(Status.IN_PROGRESS);
        }
    }, [uploadProgress, progressStatus, allFilesAttempted]);

    return (
        <div>
            <AppLayout
                contentType="form"
                content={
                    <ContentLayout
                        header={
                            <SpaceBetween size="m">
                                <Header variant="h1">Upload Documents</Header>
                            </SpaceBetween>
                        }
                        data-testid="doc-upload-view-contentlayout"
                    >
                        <form onSubmit={(event) => event.preventDefault()}>
                            <Form
                                actions={
                                    <SpaceBetween direction="horizontal" size="xs">
                                        {showProgressBar && (
                                            <ProgressBar
                                                value={uploadProgress}
                                                additionalInfo={uploadResultMessage}
                                                variant="key-value"
                                                label="Document Upload Status"
                                                status={progressStatus}
                                                data-testid="upload-document-progress-bar"
                                            />
                                        )}
                                        <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
                                        <Button
                                            data-testid="upload-documents-button"
                                            variant="primary"
                                            onClick={() => handleUploadClick()}
                                            disabled={!allowUploadFiles()}
                                        >
                                            Upload Documents
                                        </Button>
                                        <Button variant="link" onClick={() => navigate(`/`)}>
                                            Done
                                        </Button>
                                    </SpaceBetween>
                                }
                                errorIconAriaLabel="Error"
                            >
                                <Container>
                                    <SpaceBetween direction="vertical" size="l">
                                        <FormField
                                            label="Case name"
                                            i18nStrings={{ errorIconAriaLabel: 'Error' }}
                                            data-testid="case-name-field"
                                        >
                                            <Select
                                                selectedOption={selectedCaseOption}
                                                onChange={({ detail }) => setSelectedCaseOption(detail.selectedOption)}
                                                options={casesOptions}
                                                filteringType="auto"
                                                selectedAriaLabel="Selected"
                                                data-testid="upload-doc-case-select"
                                            />
                                        </FormField>
                                        <FileUpload
                                            label="Files to upload"
                                            accept=".png,.pdf,.jpg,.jpeg"
                                            fileMetadata={fileMetadata}
                                            multiple={allowMultipleFiles}
                                            value={chosenFiles}
                                            description={
                                                'Please choose the files you wish to upload to the selected case'
                                            }
                                            disabled={!allowChooseFiles()}
                                            errorText={chooseFilesError}
                                            data-testid="file-upload-field"
                                            onChange={({ detail }) => handleFileSelection(detail.value)}
                                            workflowConfig={props.workflowConfig}
                                            documentTypeList={documentTypeList}
                                            setDocumentTypeList={updateDocumentTypeList}
                                            remainingAllowedDocs={selectedCaseRemainingAllowedDocs}
                                        ></FileUpload>
                                    </SpaceBetween>
                                </Container>
                            </Form>
                        </form>
                        {uploadResultMessage === 'File upload failed' && filesFailedToUpload.length > 0 && (
                            <Alert
                                statusIconAriaLabel="Error"
                                type="error"
                                header="Failed to upload the following files: "
                            >
                                {filesFailedToUpload.map((fileName, i) => (
                                    <li key={fileName + i}>{fileName}</li> // NOSONAR - fix to make displayed file names unique to be implemented in the future
                                ))}
                            </Alert>
                        )}
                    </ContentLayout>
                }
                headerSelector="#header"
                navigationHide
                toolsHide
                date-testid="doc-upload-view-applayout"
            />
        </div>
    );
}
