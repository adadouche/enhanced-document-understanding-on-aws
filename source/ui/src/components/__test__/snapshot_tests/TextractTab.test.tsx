/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { StatusIndicatorProps } from '@cloudscape-design/components';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import { documentLines, kvPairs, tables } from '../../../__test__/test_data';
import { TEXTRACT_RAW_TEXT } from '../../../utils/constants';
import TextractTab from '../../TextractTab';

afterEach(cleanup);

const props = {
    selectedDocumentFileType: 'pdf',
    selectedDocumentUrl: 'fake-url',
    documentLines,
    kvPairs,
    tables,
    documentPageCount: 2,
    currentPageNumber: 1,
    switchPage: jest.fn(),
    textractOutputType: TEXTRACT_RAW_TEXT,
    currentStatus: 'success' as StatusIndicatorProps.Type,
    retrieveSignedUrl: jest.fn()
};

jest.mock('@cloudscape-design/components', () => {
    const Components = jest.genMockFromModule('@cloudscape-design/components') as any;
    for (const componentName of Object.keys(Components)) {
        Components[componentName] = componentName;
    }
    return Components;
});

jest.mock('react-pdf', () => ({
    pdfjs: { GlobalWorkerOptions: { workerSrc: 'abc' } },
    Document: ({ onLoadSuccess = (pdf = { numPages: 4 }) => pdf.numPages }) => {
        return <div>{onLoadSuccess({ numPages: 4 })}</div>;
    },
    Outline: null,
    Page: () => <div>def</div>
}));

test('Textract tab snapshot test', async () => {
    const tree = renderer.create(<TextractTab {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
});
