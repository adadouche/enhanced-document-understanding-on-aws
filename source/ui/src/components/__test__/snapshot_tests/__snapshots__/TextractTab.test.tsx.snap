// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textract tab snapshot test 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "height": "100%",
    }
  }
>
  <div
    style={
      Object {
        "float": "left",
        "paddingLeft": "1%",
        "paddingRight": "0.5%",
        "width": "50%",
      }
    }
  >
    <Container
      data-testid="textract-raw-text"
    >
      <div>
        <div
          className="pdf-viewer"
          data-testid="pdf-box"
        >
          <div
            className="react-pdf__Document document"
          >
            <div
              className="react-pdf__message react-pdf__message--loading"
            >
              Loading PDFâ€¦
            </div>
          </div>
          <Pagination
            ariaLabels={
              Object {
                "nextPageLabel": "Next page",
                "pageLabel": [Function],
                "previousPageLabel": "Previous page",
              }
            }
            currentPageIndex={1}
            data-testid="pdf-pagination"
            onChange={[Function]}
            pagesCount={0}
          />
        </div>
      </div>
    </Container>
  </div>
  <div
    style={
      Object {
        "float": "left",
        "height": "100%",
        "paddingLeft": "0.5%",
        "paddingRight": "1%",
        "width": "50%",
      }
    }
  >
    <Container
      fitHeight={true}
    >
      <div>
        <Box>
          <div>
            <Box
              display="inline"
              variant="h3"
            >
               
              Detected Raw Text:
               
            </Box>
            <ul
              data-testid="raw-text-data"
            >
              <ExpandableSection
                headerText="Page 1"
              >
                <ol>
                  <li
                    onClick={[Function]}
                    style={
                      Object {
                        "cursor": "pointer",
                      }
                    }
                  >
                    fake text 1
                  </li>
                  <li
                    onClick={[Function]}
                    style={
                      Object {
                        "cursor": "pointer",
                      }
                    }
                  >
                    fake text 2
                  </li>
                </ol>
              </ExpandableSection>
            </ul>
            <footer />
          </div>
        </Box>
      </div>
    </Container>
  </div>
</div>
`;
